lib/src/main/java/uk/gov/di/ipv/cri/passport/library/validation/JarValidator.java

```diff
@@ -0,0 +1,187 @@
+
+    public JarValidator(ConfigurationService configurationService) {
+        this.configurationService = configurationService;
+    }
+
+    public SignedJWT decryptJWE(String jweString) throws SharedAttributesValidationException {
+        try {
+            JWEObject jweObject = JWEObject.parse(jweString);
+
+            // Decrypt functionality to go here...
+
+            return jweObject.getPayload().toSignedJWT();
+        } catch (ParseException e) {
+            LOGGER.error("Failed to parse request body into a JWE");
+            throw new SharedAttributesValidationException(OAuth2Error.INVALID_REQUEST_OBJECT);
+        }
+    }
+
+    public JWTClaimsSet validateRequestJwt(SignedJWT signedJWT, String clientId)
+            throws SharedAttributesValidationException {
+        validateClientId(clientId);
+        validateJWTHeader(signedJWT, clientId);
+        validateSignature(signedJWT, clientId);
+        JWTClaimsSet claimsSet = validateClaimSet(signedJWT, clientId);
+        validateRedirectUri(claimsSet, clientId);
+
+        return claimsSet;
+    }
+
+    private void validateClientId(String clientId) throws SharedAttributesValidationException {
+        try {
+            configurationService.getClientAuthenticationMethod(clientId);
+        } catch (ParameterNotFoundException e) {
+            LOGGER.error("Unknown client id provided {}", clientId);
+            throw new SharedAttributesValidationException(
+                    OAuth2Error.INVALID_CLIENT.setDescription("Unknown client id was provided"));
+        }
+    }
+
+    private void validateJWTHeader(SignedJWT signedJWT, String clientId)
+            throws SharedAttributesValidationException {
+        JWSAlgorithm configuredAlgorithm =
+                JWSAlgorithm.parse(configurationService.getClientSigningAlgorithm(clientId));
+        JWSAlgorithm jwtAlgorithm = signedJWT.getHeader().getAlgorithm();
+        if (jwtAlgorithm != configuredAlgorithm) {
+            LOGGER.error(
+                    "jwt signing algorithm {} does not match signing algorithm configured for client: {}",
+                    jwtAlgorithm,
+                    configuredAlgorithm);
+            throw new SharedAttributesValidationException(
+                    OAuth2Error.INVALID_REQUEST_OBJECT.setDescription(
+                            "Signing algorithm used does not match required algorithm configured for client"));
+        }
+    }
+
+    private void validateSignature(SignedJWT signedJWT, String clientId)
+            throws SharedAttributesValidationException {
+        try {
+            SignedJWT concatSignatureJwt;
+            if (JwtHelper.signatureIsDerFormat(signedJWT)) {
+                concatSignatureJwt = JwtHelper.transcodeSignature(signedJWT);
+            } else {
+                concatSignatureJwt = signedJWT;
+            }
+            boolean valid =
+                    concatSignatureJwt.verify(
+                            new ECDSAVerifier(
```

Comment:
I notice that earlier in the process we check for the algorithm to make sure it matches that configured for the client id in the config. That implies different algorithms might be being supported. If that is the case, then this line is assuming the algorithm is always ECDS which is inconsistent with that. 

If we only support verifying signatures using ECDS that we confirm the algorithm in the header and that configured is ECDS.

Reply:
Oh yeah good shout. I got this validation check from what kbv team are doing on their address CRI but forgot about the changes that have in the signature verification bit. When retrieving the signing key from their configuration this can either be a RSAPublicKey or an ECPublicKey. Not sure why they need to do that tbh, might have just been a future proofing thing.

For us though I don't think we have an immediate need to support more than one type, or different types depending on client etc. I'll change this to be as you said above, just a set check of "Is the header signature value EC256" and then use the ECDSVerifier to validate the signature. It means we won't need that new SSM parameter for what signing algorithm we are expecting as we can just hard code that to be ES256 for all clients